fathom/src/surface/elaboration/reporting.rs

```diff
@@ -307,13 +306,18 @@ impl Message {
             Message::BooleanLiteralNotSupported { range } => Diagnostic::error()
                 .with_message("boolean literal not supported for expected type")
                 .with_labels(vec![Label::primary(file_id, *range)]),
-            Message::FailedToUnify { range, error } => {
+            Message::FailedToUnify {
+                range,
+                lhs,
+                rhs,
+                error,
+            } => {
                 use unification::{Error, RenameError, SpineError};
 
                 // TODO: Make these errors more user-friendly
                 match error {
-                    Error::Mismatch => Diagnostic::error()
-                        .with_message("type mismatch")
+                    Error::Mismatch(_, _) => Diagnostic::error()
+                        .with_message(format!("type mismatch, expected: {}, got {}", lhs, rhs))
```

Comment:
Is there any reason to store the values here, seeing as they are not used at the mometn? Was it for improved error reporting in the future? (which I can get behind, just curious)

Reply:
My understanding/reasoning is it's due to the core/surface separation. `Error::Mismatch` is created in the core language and it has to carry the values up to where the `Message` is created from it and the values pretty printed:

```
  match self.unification_context().unify(type0, type1) {
            Ok(()) => expr,
            Err(error) => {
                let lhs = self.pretty_print_value(type0);
                let rhs = self.pretty_print_value(type1);
                self.push_message(Message::FailedToUnify {
                    range,
                    lhs,
                    rhs,
                    error,
                });
                core::Term::Prim(Prim::ReportedError)
            }
        }
```

Ideally I think we'd drop the `error` field from `FailedToUnify` since it's no longer needed in this case after the values are pretty printed. However, the code at this point is matching on the error for other reasons unification can fail. Perhaps a new variant needs to be added to `Message` to handle this case separately?